Implemented simple server backups

moldfield1995 opened the pull request:
Proposed Implementation for  #1327 and #1180 

This added the ability to create, download and restore from a backup for each server. This creates a full backup of the server as a tar, allowing the user to name them and shows the size and when it was taken.


![image](https://github.com/user-attachments/assets/cab39858-ecc4-4d32-b700-89bda545e883)

The backups can be configured to be stored in a separate location from the servers if you want the backups to be on separate location, but defaults to the same directory as the nods servers.

Backups can only be taken when the server is stopped. and if a backup or restore is taking place the server can not be started or another backup action can be taken.

Known Issues:

1. Date time format for English is always in MM/DD/YYYY format (English is en-us and there is no current option for en-gb)

Happy for any feedback and change requests, I felt like this was a good stopping point and any extra features could be done in a separate PR (e.g. backups on restart or shutdown, limiting backups, and defining the compression type)

moldfield1995 committed:
Added creating and viewing backups

Added backup table to store all backups per server
Added ability to create full backups
Added backup tab on server to view and create new backups
Added backup path config
Added backup scopes

moldfield1995 committed:
Added deleting backups

Added endpoints and service for deleting backups
Added warning prompt when deleting a backup
Added loading when deleting a backup
Updated toast messages

moldfield1995 committed:
Added backup restore

Added endpoints and services for restoring a backup
Added warning prompt when restoring a backup
Updated incorrect doc endpoints

moldfield1995 committed:
Added backup download

Added endpoints and services for downloading a backup
Updated some incorrect endpoint documentation

moldfield1995 committed:
Added check when allready backing up or restoring

Added checks if a backup is already in progress when creating, restoring and deleting backup
Added new error messages
Disabled buttons when doing a blocked action in the backup screen
Added check if request is blocked because backup is in progress

moldfield1995 committed:
Added scopes to server user management

Added backup scopes to server user management
Updated backup restore to use the correct permission

moldfield1995 committed:
Added running checks

Added running checks when backing up and restoring
Added checks to start and install if a backup action is in progress

moldfield1995 committed:
Fixed build errors, Fixed restore failing, Moved to Snappy compression

Updated references to new modules
Loop through all files in root dir instead of deleting root dir
Moved compression to tar.sz (should be faster compression and extraction

moldfield1995 committed:
General Cleanup

Removed debugging logs
Cleaned up formatting
Removed old length check

LordRalex commented on servers/server.go:
```diff
@@ -601,6 +611,138 @@ func (p *Server) Extract(source, destination string) error {
+		return "", 0, pufferpanel.ErrSettingNotConfigured("backupDirectory")
+	}
+
+	if p.GetEnvironment().IsBackingUp() {
```
Technically a lock or something should be used, but this is still okay.

moldfield1995 replied:
I followed the pattern from Install, can add a lock here or leave it, happy with either

LordRalex commented on servers/server.go:
```diff
@@ -601,6 +611,138 @@ func (p *Server) Extract(source, destination string) error {
+	if err != nil {
+		return "", 0, err
+	}
+	backupFileName := backupId.String() + ".tar.sz"
```
tar.sz?

moldfield1995 replied:
its S2 snappy compression, I had a look at the compress package for the best speed vs compression. 
did this as a last change as didn't know if performance for backing up and restoring vs compatibility was best.
Can move it back to tar.gz if that is preferred?

LordRalex replied:
Fair enough, didn't see sz before. I think speed is probably better in this case.

LordRalex commented on servers/server.go:
```diff
@@ -601,6 +611,138 @@ func (p *Server) Extract(source, destination string) error {
+			p.GetFileServer().Remove(existingFile)
+		}
+		if err != nil {
+			logging.Info.Printf("failed to delted %s", err)
```
typo for deleted

moldfield1995 replied:
fixed in [3afa64c](https://github.com/pufferpanel/pufferpanel/pull/1379/commits/3afa64cabbf2a66b294418fec744d28d5024bb52)

LordRalex commented on services/backup.go:
```diff
@@ -0,0 +1,64 @@
+}
+
+func (bs *Backup) Update(model *models.Backup) error {
+	res := bs.DB.Omit(clause.Associations).Omit("id").Save(model)
```
The omitting of the id should not be needed. Gorm will skip PKs.

moldfield1995 replied:
fixed in [3afa64c](https://github.com/pufferpanel/pufferpanel/pull/1379/commits/3afa64cabbf2a66b294418fec744d28d5024bb52)

LordRalex commented on web/api/servers.go:
```diff
@@ -978,6 +989,215 @@ func editServerDataAdmin(c *gin.Context) {
+		return
+	}
+
+	c.Status(http.StatusOK)
```
If there is no body, use StatusNoContent instead

moldfield1995 replied:
fixed in [3afa64c](https://github.com/pufferpanel/pufferpanel/pull/1379/commits/3afa64cabbf2a66b294418fec744d28d5024bb52)

LordRalex commented on web/api/servers.go:
```diff
@@ -978,6 +989,215 @@ func editServerDataAdmin(c *gin.Context) {
+		return
+	}
+
+	c.Status(http.StatusOK)
```
If there is no body, use StatusNoContent instead

moldfield1995 replied:
fixed in [3afa64c](https://github.com/pufferpanel/pufferpanel/pull/1379/commits/3afa64cabbf2a66b294418fec744d28d5024bb52)

LordRalex commented on web/api/servers.go:
```diff
@@ -978,6 +989,215 @@ func editServerDataAdmin(c *gin.Context) {
+		return
+	}
+
+	c.Status(http.StatusOK)
```
If there is no body, use StatusNoContent instead

moldfield1995 replied:
fixed in [3afa64c](https://github.com/pufferpanel/pufferpanel/pull/1379/commits/3afa64cabbf2a66b294418fec744d28d5024bb52)

LordRalex commented on web/api/servers.go:
```diff
@@ -978,6 +989,215 @@ func editServerDataAdmin(c *gin.Context) {
+	newHeaders := cleanHttpReturnErrors(callResponse.Header)
+
+	c.DataFromReader(callResponse.StatusCode, callResponse.ContentLength, callResponse.Header.Get("Content-Type"), callResponse.Body, newHeaders)
+	c.Abort()
```
Should not need an abort, this should be the last op

moldfield1995 replied:
fixed in [3afa64c](https://github.com/pufferpanel/pufferpanel/pull/1379/commits/3afa64cabbf2a66b294418fec744d28d5024bb52)

LordRalex commented on web/api/servers.go:
```diff
@@ -1075,9 +1295,16 @@ func proxyHttpRequest(c *gin.Context, path string, ns *services.Node, node *mode
+	newHeaders := cleanHttpReturnErrors(callResponse.Header)
+
+	c.DataFromReader(callResponse.StatusCode, callResponse.ContentLength, callResponse.Header.Get("Content-Type"), callResponse.Body, newHeaders)
+	c.Abort()
```
Should not need an abort, this should be the last op

moldfield1995 replied:
fixed in [3afa64c](https://github.com/pufferpanel/pufferpanel/pull/1379/commits/3afa64cabbf2a66b294418fec744d28d5024bb52)

LordRalex commented on web/daemon/server.go:
```diff
@@ -815,6 +820,116 @@ func extract(c *gin.Context) {
+		response.HandleError(c, err, http.StatusBadRequest)
+	} else if response.HandleError(c, err, http.StatusInternalServerError) {
+	} else {
+		c.Status(http.StatusOK)
```
StatusNoContent

moldfield1995 replied:
fixed in [3afa64c](https://github.com/pufferpanel/pufferpanel/pull/1379/commits/3afa64cabbf2a66b294418fec744d28d5024bb52)

LordRalex commented on web/daemon/server.go:
```diff
@@ -815,6 +820,116 @@ func extract(c *gin.Context) {
+		response.HandleError(c, err, http.StatusBadRequest)
+	} else if response.HandleError(c, err, http.StatusInternalServerError) {
+	} else {
+		c.Status(http.StatusOK)
```
StatusNoContent

moldfield1995 replied:
fixed in [3afa64c](https://github.com/pufferpanel/pufferpanel/pull/1379/commits/3afa64cabbf2a66b294418fec744d28d5024bb52)

nepcore commented on client/frontend/src/components/server/Backup.vue:
```diff
@@ -0,0 +1,178 @@
+    )
+}
+
+async function deleteRestore(file) {
```
`deleteRestore`?

moldfield1995 replied:
something tells me I was working on that late, updated in [9eadb8a](https://github.com/pufferpanel/pufferpanel/pull/1379/commits/9eadb8aef397ec2bca7c4bf830696dc462390075) to deleteBackup

moldfield1995 committed:
Code standard and spelling fixes

Updated spelling mistakes
Removed unneeded Omit from backup update
Updated StatusOk to StatusNoContent where body is

nepcore commented on client/frontend/src/components/server/Backup.vue:
```diff
@@ -0,0 +1,178 @@
+  }
+}
+console.log(locale.value.replace('_', '-'))
+// date-fns localization approach takes importing an object per locale
+// which is both a lot of data and annoying to have to handle as that
+// means manually dealing with fallbacks, dynamic imports, etc
+// so instead we just force the adapter to use the browsers built in
+// date localization tooling and just let date-fns do the logic parts
```
copy/pasted comment that doesn't apply here

moldfield1995 replied:
Removed in [9eadb8a](https://github.com/pufferpanel/pufferpanel/pull/1379/commits/9eadb8aef397ec2bca7c4bf830696dc462390075)

nepcore commented on client/frontend/src/components/server/Backup.vue:
```diff
@@ -0,0 +1,178 @@
+    loading.value = false
+  }
+}
+console.log(locale.value.replace('_', '-'))
```
i'd argue this is probably not the right place to log the locale beyond debugging :wink:

moldfield1995 replied:
ah forgot to remove that, was having some local bugs with locale

moldfield1995 replied:
Removed in [9eadb8a](https://github.com/pufferpanel/pufferpanel/pull/1379/commits/9eadb8aef397ec2bca7c4bf830696dc462390075)

moldfield1995 committed:
Updated backup vue

Removed debug and unnecessary comments
Updated delete method name to make more sense

LordRalex commented on web/daemon/server.go:
```diff
@@ -815,6 +820,116 @@ func extract(c *gin.Context) {
+	fileName := c.Query("fileName")
+	isRunning, err := server.IsRunning()
+	if err != nil {
+		response.HandleError(c, err, http.StatusInternalServerError)
```
response.HandleError actually does a if err != nil, so the check isn't needed.

You can just do:
```go
if response.HandleError(c, err, http.StatusInternalServerError) {
  return
}
```

moldfield1995 replied:
Fixed the above in [8208ec1](https://github.com/pufferpanel/pufferpanel/pull/1379/commits/8208ec1b106b1ef7ee619f45e9ce0d915e4309d7). 
Found similar duplicate checks in web/api that I have made, so fixed them as well

moldfield1995 committed:
Removed extra nil checks

Removed duplicate error != nil checks
Cleaned up code for backupId cast

LordRalex commented on client/frontend/src/lang/en_US/backup.json:
```diff
@@ -0,0 +1,17 @@
+    "Delete": "Delete",
+    "Deleted":"Backup deleted successfully",
+    "DeletePrompt":"Permanently delete backup",
+    "DeletePromptBody":"Do you really want to delete the backup permanently? "
```
Extra space at the end

LordRalex commented on client/frontend/src/lang/en_US/backup.json:
```diff
@@ -0,0 +1,17 @@
+    "Deleted":"Backup deleted successfully",
+    "DeletePrompt":"Permanently delete backup",
+    "DeletePromptBody":"Do you really want to delete the backup permanently? "
+
```
Empty line

LordRalex commented on client/frontend/src/lang/en_US/errors.json:
```diff
@@ -45,5 +45,7 @@
   "ErrUnsupportedArch": "Architecture ${actual} not supported. Supported Architectures: ${expected}",
-  "ErrNoCommand": "No command could be determined to be used for this server"
+  "ErrNoCommand": "No command could be determined to be used for this server",
+  "ErrBackupInProgress": "The server is currently performing a backup or restore. Please wait for this process to be complete",
```
"process to complete" is probably better wording

LordRalex commented on config/entries.go:
```diff
@@ -41,6 +41,7 @@ var ClientId = asString("daemon.auth.clientId", "")
 var ClientSecret = asString("daemon.auth.clientSecret", "")
 var CacheFolder = asDataFolder("daemon.data.cache", "cache")
 var ServersFolder = asDataFolder("daemon.data.servers", "servers")
+var BackupsFolder = asDataFolder("daemon.data.backups", "backups")
```
Thinking to the future, I think we probably should have this be an object instead... because eventually it'd be "add s3" and such.

Can we have this be daemon.data.backups.folder to prep for that.

LordRalex commented on services/backup.go:
```diff
@@ -0,0 +1,64 @@
+	DB *gorm.DB
+}
+
+func (bs *Backup) GetAllBakcupsForServer(serverID string) ([]*models.Backup, error) {
```
Bakcups

LordRalex commented on web/api/servers.go:
```diff
@@ -110,6 +110,17 @@ func registerServers(g *gin.RouterGroup) {
 	g.POST("/:serverId/extract/*filename", middleware.RequiresPermission(scopes.ScopeServerFileEdit), middleware.ResolveServerPanel, proxyServerRequest)
 	g.OPTIONS("/:serverId/extract/*filename", response.CreateOptions("POST"))
 
+	g.GET("/:serverId/backup", middleware.RequiresPermission(scopes.ScopeServerBackupView), middleware.ResolveServerPanel, getBackups)
```
I think we need an endpoint to get specific information on a backup. This does change routes though

moldfield1995 replied:
as in an individual backup?
I think that something like /:serverId/backup/:backupId would be sensible for it but did not see the point for it at the moment, as the data for that backups is rather light and couldn't see what it would add at the moment over the get all one

LordRalex replied:
Yes, although I agree for now on the information being lacking. The only thing I can think of is really the "date" the backup was made.

moldfield1995 committed:
Fixed fromatting and spelling mistakes

Removed extra whitespace
Fixed spelling mistakes
Fixed backup error wording

moldfield1995 committed:
Added Get Backup Endpoint

Added Get Backup Endpoint
Fixed edge case where a backup for another server could be passed and considered valid
Added Filename in returned Json

moldfield1995 commented:
Fixed all above in [53eaf96](https://github.com/pufferpanel/pufferpanel/pull/1379/commits/53eaf96580945e6f5a6efa6e51fb252b3f47f806) and [54e81dc](https://github.com/pufferpanel/pufferpanel/pull/1379/commits/54e81dc7564a15397ed0b58d95068f47538260ed)
